bundles/org.openhab.binding.fsinternetradio/src/main/java/org/openhab/binding/fsinternetradio/internal/FSInternetRadioDiscoveryParticipant.java

```diff
@@ -146,44 +157,50 @@ public DiscoveryResult createResult(RemoteDevice device) {
                 properties.put(CONFIG_PROPERTY_IP, ip);
 
                 // add manufacturer and model, if provided
-                final String manufacturer = getManufacturer(device);
+                final @Nullable String manufacturer = getManufacturer(device);
                 if (manufacturer != null) {
                     properties.put(PROPERTY_MANUFACTURER, manufacturer);
                 }
-                final String model = getModel(device);
+                final @Nullable String model = getModel(device) != null ? getModel(device) : getFriendlyName(device);
                 if (model != null) {
                     properties.put(PROPERTY_MODEL, model);
                 }
-
-                final DiscoveryResult result = DiscoveryResultBuilder.create(uid).withProperties(properties)
-                        .withLabel(device.getDisplayString()).build();
-                return result;
+                final String thingName = (manufacturer == null) && (getModel(device) == null) ? getFriendlyName(device)
+                        : device.getDisplayString();
+                return DiscoveryResultBuilder.create(uid).withProperties(properties).withLabel(thingName).build();
             }
         }
         return null;
     }
 
-    private String getManufacturer(RemoteDevice device) {
+    private @Nullable String getManufacturer(RemoteDevice device) {
```

Comment:
```
private Optional<String> getManufacturer(RemoteDevice device) {
    return Optional.ofNullable(device.getDetails()).map(DeviceDetails::getManufacturerDetails)
        .map(details -> details.getManufacturer().trim()).filter(manufacturer -> !manufacturer.isEmpty());
}
```

if `getDetails` returns an `Optional<DeviceDetails>`:

```
private Optional<String> getManufacturer(RemoteDevice device) {
    return device.getDetails().map(DeviceDetails::getManufacturerDetails)
        .map(details -> details.getManufacturer().trim()).filter(manufacturer -> !manufacturer.isEmpty());
}
```

or if `device` itself is an `Optional`

```
private Optional<String> getManufacturer(Optional<RemoteDevice> device) {
    return device.map(RemoteDevice::getDetails).map(DeviceDetails::getManufacturerDetails)
        .map(details -> details.getManufacturer().trim()).filter(manufacturer -> !manufacturer.isEmpty());
}
```

If you need the string trim-and-check-for-empty more often:

```
private Optional<String> trimAndCheckEmpty(String string) {
    return Optional.of(string).map(String::trim).filter(s -> !s.isEmpty());
}

private Optional<String> getManufacturer(Optional<RemoteDevice> device) {
    return device.map(RemoteDevice::getDetails).map(DeviceDetails::getManufacturerDetails)
        .map(ManufacturerDetails::getManufacturer).flatMap(this::trimAndCheckEmpty);
}
```

So basically: the argument of `map` is applied to the result of the previous stage if the optional is non-empty and skipped if empty. `flatMap` does the same, but (as our `trimAndCheckEmpty` returns an `Optional`) prevents nested `Optional` (otherwise we would get an `Optional<Optional<String>>`).

Reply:
ok, save this link here. This requires more time, I'll come back to that

Reply:
Ok, I did some first adaptions, could you please check the DiscoveryParticipant (every usage of Optional). I wouldn't complete that.

One questions: From my understanding the objective would be to avoid usage of @Nullable even the class is declared with @NonNullByDefault, corret. How to deal with the fact that 

`
@Nullable String s = null;
if (xxx) {
  s = "ha ha":
}
return Optional.ofNullable(s)
`

the Optional.ofNullable(s) is not accepted, because no method accepting a nullable object is defined.

If the changes are ok I stop doing the rest (significant effort and risk to get logic bugs) and you could finish the review. This requires more time.